### PR TITLE
Exploration and additional test cases

### PR DESCRIPTION
**Pull request by thebioengineer**
Hey @elimillera, I played around with your branch and made some additions that hopefully you'll like. 
- I tried to simplify the template build process, changed out to use `file.path()` as opposed to pasting manually, and set the functions to open the created file on completion. Also, now they won't override existing files. Also now the user can manually pass a name they want in the spec as opposed to us always guessing from the username (still the default)
- some renaming of functions: ie vt_use_case -> vt_use_test_case, 
- I played with a test package and id'd some bugs in the build process, so I made some edits. Main one is that the vignette process assumes the package is installed. but if building, then the package would not necessarily be installed and you could accidentally be using an old version of the package. now it is installed in a temporary library and that is the one used.
- on above I went down a path of creating a rendering function for the validation report. Now we can use the same function for rendering the validation report from either the build or installed packages.
- On the validation of an installed package, now the output file will be open in addition to being saved
- I generally added some tests

Thank you so much for doing all that work, it was great!

take a look and we can discuss edits.

I think a template for the validation report creation is still missing, and some other tooling, but I think with this addition, we are getting a good chunk closer to a point where we could solicit feedback from other users!

### CODE REVIEW

**Comment by elimillera** [comment]
@thebioengineer I really like the changes. The opening of the file after `vt_use_<x>` is a really good touch. I think the file paths were getting a little clunky and all of your changes look good to me.

**Review by elimillera — APPROVED** [review]

File: R/spec.R [hunk 1]
```diff
@@ -1,26 +1,48 @@
...
-#'   are located at `./inst/validation/specs/{spec_name}`.
+#'   are located at `./inst/validation/specification/{spec}`.
+#'
+#' @return Path to the newly created specification file, invisibly.
 #'
 #' @export
 #'
 #' @rdname new_item
-vt_use_spec <- function(spec_name, spec_content, pkg = "."){
+vt_use_spec <- function(name, username = vt_username(), open = interactive(), pkg = "."){
```
[the remark this answers is not in the document]
Comment (mariev) [reply R/spec.R:13]: Do we want to enforce good file naming practices here? No spaces, etc.

File: R/spec.R [hunk 2]
```diff
@@ -1,26 +1,48 @@
...
-  ))
-
-  writeLines(content, con = spec_path)
+  spec_name <- create_item(pkg, "specification", name)
+
+  ## if the file didnt exist before, populate with contents
+  if (file.size(spec_name) == 0){
+
+    # Create the content to write
+    content <- paste0(c(
```
[the remark this answers is not in the document]
Comment (mariev) [reply R/spec.R:25]: missing @title

File: R/case.R [hunk 3]
```diff
@@ -1,37 +1,60 @@
...
-    collapse = "",
-    sep = "\n"
-  ))
-
-  writeLines(content, con = case_path)
+#' vt_use_test_case("regTests/Update2/spec2", package_dir)
+#' }
+#' @importFrom usethis edit_file
+vt_use_test_case <- function(name, username = vt_username(), open = interactive(), pkg = ".") {
+
```
Comment (mariev) [inline R/case.R:27]: same as for `vt_use_spec` - file naming conventions?

File: R/case.R [hunk 4]
```diff
@@ -1,37 +1,60 @@
...
+  name <- vt_set_ext(name, ext = "md")
+
+  case_name <- create_item(pkg, "test_case", name)
+
+  ## if the file didnt exist before, populate with contents
+  if (file.size(case_name) == 0){
+
+    # Create the content to write
+    content <- paste0(c(
+      "#' @section Last Updated By:",
```
Comment (mariev) [inline R/case.R:37]: same as for `vt_use_spec`:

- will need to update due to #16 
- missing @title field

File: R/code.R [hunk 5]
```diff
@@ -1,14 +1,47 @@
...
+  # Create file to write in
+  code_name <- create_item(pkg, "test_code", code_name)
+
+  ## if the file didnt exist before, populate with contents
+  if (file.size(code_name) == 0){
+
+    # Create the content to write
+    content <- paste0(c(
+      "",
+      "# Test setup",
```
Comment (mariev) [inline R/code.R:23]: `context()`

File: R/init.R [hunk 6]
```diff
@@ -12,13 +12,12 @@
 #' @export
 vt_use_validation <- function(pkg = ".") {
 
+  validation_directory <- getOption("vt.validation_directory",default = "vignettes/validation")
```
Comment (mariev) [inline R/init.R:15]: can we have $get and $set access for future extensibility? similar to `knitr::opts_knit`

File: R/utils.R [hunk 7]
```diff
@@ -1,56 +1,102 @@
...
-                         sep = .Platform$file.sep, collapse = "")
+  item_dir <- dirname(item_name)
+
   # If the item is nested in a folder make sure its made.
-  if(length(item_name_split) > 1){
+  if( !item_dir %in% c("",".")){
     tryCatch({
-      dir.create(val_dir_path, recursive  = TRUE)
-
+      dir.create(file.path(pkg,validation_directory,type,item_dir), recursive  = TRUE, showWarnings = FALSE)
```
Comment (mariev) [inline R/utils.R:36]: use `fs::dir_create` to silently ignore existing directories

File: R/utils.R [hunk 8]
```diff
@@ -1,56 +1,102 @@
...
   }
 
-  item_file_path <- paste0(c(val_dir_path, .Platform$file.sep,tail(item_name_split, 1)),
-                           collapse = "")
+  item_file_path <- file.path(pkg,validation_directory, type, item_name)
 
-  file.create(item_file_path)
+  tryCatch({
+
+    file.create(item_file_path)
```
Comment (mariev) [inline R/utils.R:49]: `fs::file_create`

**Review by mariev — CHANGES_REQUESTED** [review]
General comments:

- `vt_use_*` will need refactoring once #16 is implemented, not blocking for now
- Wondering how we expect these templates to play with `bookdown`? Having naked roxygen blocks in the child Rmd files can cause problems, although this can be circumvented by using yaml header around the roxygen chunk
- I'd like to see us tighten up language around requirements and specifications. A specification is a set of requirements, where each of the files generated via `vt_use_spec` is a single requirement. Presumably an org could circulate a spec document for approvals/review prior to engaging in the work of writing test cases or test code.
- How should users address the use case of preparing code that needs to be available both during install and post-install for validation?  Does subscribing to the "validate on install" model preclude the "validate post install" or perhaps can we build it into `vt_validated_build`? Trying to reduce the lift when switching perspectives.
- How should we consider the intersection of validation reports with  `usethis::use_vignette` for other vignettes? This action will add .gitignore file to vignette folder.


Coding conventions:
- why does `vt_use_validation` create the validation structure and not start a validation vignette? I'd like to see this renamed `vt_create_validation`, as a function that allows users to add validation structure to an existing location, and a new function `vt_use_validation` to mean initiate code via a validation report template. 
-  Some further consideration around `vt_create_*` vs `vt_use_*` functions: when are we creating folder structures and when are we just adding files? 🤔 Can `vt_create_validation` add all the folders while `vt_use_*` add only the folder necessary for the new file?
- can `vt_path` and `vt_username` be pulled into an `options` function?

Style question:
- Why are we using `writeLines` for content generation? imo we should be using templating features similar to VISCtemplates

File: R/code.R (hunk 5 above)
Comment (thebioengineer) [reply R/code.R:23]: `context()` is [deprecated in the third edition](https://testthat.r-lib.org/reference/context.html) and it sounds like generally its use is discouraged now. Do we use `context()` for anything in our current code?

File: R/init.R (hunk 6 above)
Comment (thebioengineer) [reply R/init.R:15]: absolutely. I was getting a little concerned with the scope of the PR, so was waiting on that.

Comment (mariev) [reply R/init.R:15]: yeah, let's just open a new issue for this

File: R/code.R (hunk 5 above)
Comment (mariev) [reply R/code.R:23]: gtk. our legacy scrape function expects it 😢, luckily we haven't implemented anything in {valtools} yet...

**Comment by thebioengineer** [comment]
> * `vt_use_*` will need refactoring once #16 is implemented, not blocking for now

we can just implement those changes here before merging this branch (#17)

> * Wondering how we expect these templates to play with `bookdown`? Having naked roxygen blocks in the child Rmd files can cause problems, although this can be circumvented by using yaml header around the roxygen chunk

I think we originally looked at bookdown to facilitate the dynamic referencing, correct? do we want to continue, or intend that our markdown files require "preprocessing" to be able to be used in the context of normal markdowns

> * How should users address the use case of preparing code that needs to be available both during install and post-install for validation?  Does subscribing to the "validate on install" model preclude the "validate post install" or perhaps can we build it into `vt_validated_build`? Trying to reduce the lift when switching perspectives.

My thought was that the `vt_path()` function would facilitate that. Any content necessary for the validation lives in `vignettes/validation` (or whatever folder they designate). that contents are copied into `inst/validation` via `vt_validated_build()`. Then on install, to rerun the validation, they use `vt_validate_installed_package()`. `vt_path()` uses references that update based on the call.

> * How should we consider the intersection of validation reports with  `usethis::use_vignette` for other vignettes? This action will add .gitignore file to vignette folder.

The .gitignore is of ,R and and .html files, but definitely makes things more difficult. I assumed we would implement our own version, like `vt_use_report()`, that would actually be called by the setup functions, but that doesn't resolve it when the users still call `usethis::use_vignette()`.

> 
> Coding conventions:
> 
> * why does `vt_use_validation` create the validation structure and not start a validation vignette? I'd like to see this renamed `vt_create_validation`, as a function that allows users to add validation structure to an existing location, and a new function `vt_use_validation` to mean initiate code via a validation report template.

I think it would make sense to implement that feature!

> * Some further consideration around `vt_create_*` vs `vt_use_*` functions: when are we creating folder structures and when are we just adding files? 🤔 Can `vt_create_validation` add all the folders while `vt_use_*` add only the folder necessary for the new file?

Also reasonable. However we should still have `vt_use_*` create the folders if they don't exist b/c git doesnt track folders unless they have content~

> * can `vt_path` and `vt_username` be pulled into an `options` function?

Could you elaborate? as in be able to set the values?

File: R/spec.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (thebioengineer) [reply R/spec.R:13]: Updated to enforce good file name practices (cribbed/coped from the un-exported `usethis:::check_file_name()` function.

File: R/spec.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (thebioengineer) [reply R/spec.R:25]: > will need to be updated in response to #16, not blocking for now

updated to using editor/editDate, and capturing title

File: R/spec.R [hunk 9]
```diff
@@ -1,26 +1,48 @@
...
+      "#' @section Last Updated By:",
+      paste0("#' ", username),
+      "#' @section Last Update Date:",
+      paste0("#' ", as.character(Sys.Date())),
+      "",
+      "+ _Specifications_",
+      "    + Start documenting specifications here!"
```
[the remark this answers is not in the document]
Comment (thebioengineer) [reply R/spec.R:32]: Added wording and content to be:

```
#' @title TITLE AS PASSED
#' @editor USERNAME
#' @editDate YYYY-MM-DD

+ Start documenting requirements here!
```

File: R/case.R (hunk 3 above)
Comment (thebioengineer) [reply R/case.R:27]: see above

File: R/spec.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (elimillera) [reply R/spec.R]: @mariev I would agree with your change here. Once this is pulled in I can change this to `vt_use_req()` and change references to specifications to requirements.

**Comment by elimillera** [comment]
It looks like Ellis got most of your comments but are a couple of my thoughts

> I'd like to see us tighten up language around requirements and specifications. A specification is a set of requirements, where each of the files generated via vt_use_spec is a single requirement. Presumably an org could circulate a spec document for approvals/review prior to engaging in the work of writing test cases or test code.

Totally agree. I think once I can get the naming convention right for this PR it should stick for the rest of the package 😅. I think the functionality for rendering a single piece of the validation document for signatures/approval is some low hanging fruit for us.

> Why are we using writeLines for content generation? imo we should be using templating features similar to VISCtemplates

`usethis::use_template` has been added to my package dev repertoire. 👍 I can add that in and update my PR once merging is done.

**Comment by thebioengineer** [comment]
The downside I can see to using `usethis::use_template` is that it expects you to using RProjects. Granted, most people are, but that can cause a sticking point for folks that are against that workflow

File: R/spec.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (mariev) [reply R/spec.R]: see #20

File: R/init.R (hunk 6 above)
Comment (mariev) [reply R/init.R:15]: see #19

**Comment by mariev** [comment]
@thebioengineer it should be possible to get around the Rproject expectation in `use_template` by setting .here appropriately via `here::set_here()`

**Review by mariev — APPROVED** [review]
I believe the consensus is that this branch is already quite messy and we need to address new issues via separate branches. Unresolved issues spawned from  this PR include:

- #23 
- #19 
- #20 
- #24

**Comment by elimillera** [comment]
@mariev Agreed. I'll merge this into my development branch and we can pick up the PR there.

**Pull request merged**